server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java

```diff
@@ -116,11 +117,12 @@ public CompletableFuture<R> consumeIteratorAndExecute() {
     }
 
     private void continueConsumptionOrFinish(@Nullable R result, Throwable failure) {
+        if (earlyTerminationCondition.apply(result, failure)) {
+            setResult(null, resultsToFailure.apply(result, failure));
+            return;
+        }
+
         if (result != null) {
-            if (earlyTerminationCondition.test(result)) {
-                setResult(null, resultsToFailure.apply(result));
-                return;
-            }
             resultRef.accumulateAndGet(result, combiner);
```

Comment:
for fail fast on insert we use short-circuit mechanism via propagating `shardResponse.failure`, so `ShardResponseActionListener` can complete with an exception 
and `continueConsumptionOrFinish` receives args (null, throwable). 

Hence, moved the check out of the `if (result != null)`


BTW, we have a code branch on master, that hits `(null, maybe-not-null)`
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java#L197

I wonder if it means that `COPY FROM` with fail_fast = true can skip fail-fast logic for that scenario.

Reply:
Looks like `shardResponse.failure` is passed via `UpsertResults` not via the `failure` parameter of `continueConsumptionOrFinish`. So I do not think we need to move the code block out of the `if` stmt.


> BTW, we have a code branch on master, that hits (null, maybe-not-null)
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java#L197
I wonder if it means that COPY FROM with fail_fast = true can skip fail-fast logic for that scenario

The scenario here is when an exception is thrown during loading next batch? I think you are right, that we should terminate early here. (just wondering why we never terminated in such case before, even before earlyTermination was added)

Reply:
> So I do not think we need to move the code block out of the if stmt.

it won't work if I do it, test will again insert almost 60000 records. 

>Looks like shardResponse.failure is passed via UpsertResults

No, see comment below

>I initially started with modifying RowCountCollector.processShardResponse to account for that standalone failure but, as mentioned above (https://github.com/crate/crate/pull/17689#discussion_r2035155167), with "short-circuit error" we complete future exceptionally and hence continueConsumptionOrFinish receives null-result and not-null throwable on failure.